Sources/NIOCore/AsyncSequences/NIOAsyncWriter.swift

```diff
@@ -0,0 +1,750 @@
+    /// Otherwise, the call to ``NIOAsyncWriter/yield(contentsOf:)`` will get suspended until the ``NIOAsyncWriter``
+    /// becomes writable again.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(contentsOf:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// This can be called more than once and from multiple `Task`s at the same time.
+    ///
+    /// - Parameter contentsOf: The sequence to yield.
+    @inlinable
+    public func yield<S: Sequence>(contentsOf sequence: S) async throws where S.Element == Element {
+        try await self.storage.yield(contentsOf: sequence)
+    }
+
+    /// Yields an element to the ``NIOAsyncWriter``.
+    ///
+    /// If the ``NIOAsyncWriter`` is writable the element will get forwarded to the ``NIOAsyncWriterDelegate`` immediately.
+    /// Otherwise, the call to ``NIOAsyncWriter/yield(contentsOf:)`` will get suspended until the ``NIOAsyncWriter``
+    /// becomes writable again.
+    ///
+    /// If the ``NIOAsyncWriter`` is finished while a call to ``NIOAsyncWriter/yield(contentsOf:)`` is suspended the
+    /// call will throw a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// This can be called more than once and from multiple `Task`s at the same time.
+    ///
+    /// - Parameter element: The element to yield.
+    @inlinable
+    public func yield(_ element: Element) async throws {
+        try await self.yield(contentsOf: CollectionOfOne(element))
+    }
+
+    /// Finishes the writer.
+    ///
+    /// Calling this function signals the writer that any suspended or subsequent calls to ``NIOAsyncWriter/yield(contentsOf:)``
+    /// or ``NIOAsyncWriter/yield(_:)`` will return a ``NIOAsyncWriterError/alreadyFinished`` error.
+    ///
+    /// - Note: Calling this function more than once has no effect.
+    /// - Parameter completion: A ``NIOAsyncWriterCompletion`` which indicates whether publishing has finished normally
+    /// or failed with an error.
+    @inlinable
+    public func finish(completion: NIOAsyncWriterCompletion<Failure>) {
+        self.storage.finish(completion: completion)
+    }
+}
+
+@available(macOS 10.15, iOS 13, tvOS 13, watchOS 6, *)
+extension NIOAsyncWriter {
+    /// The underlying sink of the ``NIOAsyncWriter``. This type allows to toggle the writability of the ``NIOAsyncWriter``.
+    public struct Sink {
+        @usableFromInline
+        /* private */ internal var storage: Storage
+
+        @inlinable
+        init(storage: Storage) {
+            self.storage = storage
+        }
+
+        /// Toggles the writability of the ``NIOAsyncWriter``.
+        ///
+        /// If the writer becomes writable again all suspended yields will be resumed and the produced elements will be forwarded via
+        /// the ``NIOAsyncWriterDelegate/didYield(contentsOf:)`` method.
+        public func toggleWritability() {
```

Comment:
Is this the right API? I wonder if setting the writability explicitly would be more useful (if anything it makes the tests more obvious). In `actor` versions it _had_ to be toggle because calls to set the writability could be reordered.

Reply:
I kept this in because to get the writability of a `Channel` you need to load the value from the underlying atomic. Since, that load is normally done using a `relaxed` load I wasn't sure if this is safe to do in the `channelWritabilityChanged` method. However, after talking with @Lukasa this is safe to do so I am changed the API to be `setWritability(to:)`.